lib/units/api/controllers/user.js

```diff
@@ -400,3 +401,56 @@ function getUserAccessTokens(req, res) {
       })
     })
 }
+
+function addAdbPublicKey(req, res) {
+  var data = req.swagger.params.adb.value
+  adbkit.util.parsePublicKey(data.publickey)
+    .then(function(key) {
+      return dbapi.lookupUsersByAdbKey(key.fingerprint)
+        .then(function(cursor) {
+          return cursor.toArray()
+        })
+        .then(function(users) {
+          return {
+            key: {
+              title: data.title || key.comment
+            , fingerprint: key.fingerprint
+            }
+          , users: users
+          }
+        })
+    })
+    .then(function(data) {
+      if (data.users.length) {
+        return res.json({
+          success: true
+        })
+      }
+      else {
+        return dbapi.insertUserAdbKey(req.user.email, data.key)
+          .then(function() {
+            return res.json({
+              success: true
+            })
+          })
+      }
+    })
+    .then(function() {
+      req.options.push.send([
+        req.user.group
+      , wireutil.envelope(new wire.AdbKeysUpdatedMessage())
+      ])
+    })
+    .catch(dbapi.DuplicateSecondaryIndexError, function() {
+      // No-op
+      return res.json({
+        success: true
+      })
+    }).catch(function(err) {
+      log.error('Failed to insert new adb key fingerprint: ', err.stack)
+      return res.status(500).json({
```

Comment:
Shouldn't we distinguish cases when supplied key is malformed and respond with error 400 in such case?

Reply:
adbkit does not provide a specific exception object we can catch. This is what adbkit do when encountering such input:
```
reject new Error "Unrecognizable public key format"
```

What we can do better here probably is passing through the error message. Would that be suffice?

Reply:
@koral-- , is it possible now to merge soon this interesting feature ?

Reply:
Sorry for the delay.
LGTM

> What we can do better here probably is passing through the error message. Would that be suffice?

Stacktrace from that error is printed to the log so I think it is OK to leave it as is.